File: product-service/src/main/java/com/intexsoft/webshop/productservice/model/Subcategory.java
package com.intexsoft.webshop.productservice.model;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;
import org.hibernate.proxy.HibernateProxy;

import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "subcategory")
@Getter
@Setter
@ToString
@NoArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Subcategory {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Long id;
    @Column(name = "name", nullable = false, unique = true)
    String name;
    @Column(name = "description", nullable = false)
    String description;
    @ManyToOne(fetch = FetchType.LAZY)
    @ToString.Exclude
    Category category;
    @OneToMany(
            mappedBy = "subcategory",
            fetch = FetchType.LAZY,
            cascade = CascadeType.ALL
    )
    @ToString.Exclude
    Set<Attribute> attributes = new LinkedHashSet<>();

    @OneToMany(
            mappedBy = "subcategory",
            fetch = FetchType.LAZY
    )
    @ToString.Exclude
    Set<Product> products = new LinkedHashSet<>();

    public void addAttribute(Attribute attribute) {
        attributes.add(attribute);
        attribute.setSubcategory(this);
    }

    @Override
    public final boolean equals(Object object) {
        if (this == object) return true;
        if (object == null) return false;
        Class<?> oEffectiveClass = object instanceof HibernateProxy ? ((HibernateProxy) object).getHibernateLazyInitializer().getPersistentClass() : object.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        Subcategory that = (Subcategory) object;
        return getId() != null && Objects.equals(getId(), that.getId());
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass().hashCode() : getClass().hashCode();
    }
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/model/Subcategory.kt
package com.intexsoft.webshop.productservicekt.model;

import jakarta.persistence.*

@Entity
@Table(name = "subcategory")
class Subcategory(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    @Column(name = "name", nullable = false, unique = true)
    var name: String,
    @Column(name = "description", nullable = false)
    var description: String,
    @ManyToOne(fetch = FetchType.LAZY)
    var category: Category,
    @OneToMany(mappedBy = "subcategory", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var attributes: MutableList<Attribute>?,
    @OneToMany(mappedBy = "subcategory", fetch = FetchType.LAZY)
    var products: MutableList<Product>?
) {
    fun addAttribute(attribute: Attribute) {
        if (attributes == null) attributes = mutableListOf()
        attributes!!.add(attribute)
        attribute.subcategory = this
    }
}